File: coderun/2. Cheapest way (Самый дешевый путь)/dp, store row/dp_store_row.py
N, M = map(int, input().split(" "))

dp = [0] * (M)

#print("echo dp: ", dp)

inf = 3901

# first row
row = list(map(int, input().split(" ")))
#print("echo row: ", row)
for j in range(M):
    from_left = inf
    if j > 0:
        from_left = dp[j - 1]
    best = from_left
    if best == inf:
        best = 0
    #print("j:", j)
    dp[j] = best + row[j]

for i in range(N-1):
    row = list(map(int, input().split(" ")))
    for j in range(M):
        from_left, from_up = inf, inf
        from_up = dp[j]
        if j > 0:
            from_left = dp[j - 1]
        best = min(from_left, from_up)
        if best == inf:
            best = 0
        dp[j] = best + row[j]

print(dp[M - 1])

File: coderun/2. Cheapest way (Самый дешевый путь)/dp, store row/dp_store_row.cpp
#include <iostream>

using namespace std;

int main(int argc, char *argv[])
{
    int N, M;
    cin >> N >> M;

    int inf;
    inf = 3901;

    int* dp = new int[N];
    for (int j = 0; j < M; j++){
            int num;
            cin >> num;
            int fromLeft; 
            fromLeft = inf;
            if (j > 0) fromLeft = dp[j-1];
            int best = fromLeft;
            if (best == inf) best = 0;
            dp[j] = num + best;
    }

    for (int i = 1; i < N; i++){
        for (int j = 0; j < M; j++){
            int num;
            cin >> num;
            int fromLeft, fromAbove; 
            fromLeft = inf;
            fromAbove = inf;
            if (i > 0) fromAbove = dp[j];
            if (j > 0) fromLeft = dp[j-1];
            int best = min(fromAbove, fromLeft);
            if (best == inf) best = 0;
            dp[j] = num + best;
        }
    }

    cout << dp[M-1];

    return 0;
}